Add Preference to disable runtime invalidation

topolarity opened the pull request:
More conservative version of https://github.com/JuliaSIMD/HostCPUFeatures.jl/pull/21 (thanks to @el-oso for the inspiration).

Adds three preferences:
  - `cpu_target`: if provided, use this for feature detection instead of `JULIA_CPU_TARGET`
  - `freeze_cpu_target`: if `true`, "freeze" the features detected based on your precompile-time CPU target
  - `allow_runtime_invalidation`: if `false`, warn instead of invalidating when CPU features don't match precompile-time

```julia
┌ Warning: Runtime invalidation was disabled, but the CPU info is out-of-date.
│ Will continue with incorrect CPU name (from build time).
└ @ HostCPUFeatures ~/repos/HostCPUFeatures.jl/src/HostCPUFeatures.jl:62
```

This makes the package `--trim` compatible and also provides a nice preference for users who do not wish for HostCPUFeatures.jl to ever cause an "invalidation storm", even if it means running with the wrong CPU info.

topolarity committed:
Add Preference to disable runtime invalidation

This makes the package (nearly) `--trim` compatible and also provides a
nice preference for users who do not wish for HostCPUFeatures.jl to ever
cause an "invalidation storm", even if it means running with the wrong
CPU info.

oscardssmith commented on src/cpu_info_x86.jl:
```diff
@@ -32,14 +32,22 @@ fast_int64_to_double() = has_feature(Val(:x86_64_avx512dq))
 fast_half() = False()
 
-@noinline function setfeaturefalse(s)
+@inline function setfeaturefalse(s)
```
why the inlining switch?

topolarity replied:
This was to improve inferrability - `has_feature(Val(s))` is only inferrable if the literal value of `s` is available in the function, which is true in all of the callers (the argument is always a literal Symbol)

tz-lom commented on src/cpu_info.jl:
```diff
@@ -48,8 +48,12 @@ function reset_features!()
+                @debug "Defining $(has ? "presence" : "absense") of feature $feature."
+                set_feature(feature, has)
+            else
+                @warn "Runtime invalidation was disabled, but the CPU info is out-of-date.\nWill continue with incorrect CPU feature flag: $ext."
```
aside of this change, I expect that `feature_string` not always return same set of features, that will make certain features stuck in wrong state just because LLVM does not report them for that CPU

For the reference, that's the info I've collected from real and virtual machines with 
```
@ccall jl_get_cpu_name()::String
@ccall jl_get_cpu_features()::String
```

```
* "broadwell"
  "+prfchw,+avx,+aes,+sahf,+pclmul,+crc32,+sse4.1,+xsave,+sse4.2,+invpcid,+64bit,+cmov,+movbe,+rtm,+adx,+avx2,+bmi,+sse,+xsaveopt,+rdrnd,+cx8,+sse3,+fsgsbase,+lzcnt,+ssse3,+cx16,+bmi2,+fma,+popcnt,+f16c,+mmx,+sse2,+rdseed,+fxsr,-cldemote,-xop,-xsaves,-avx512fp16,-usermsr,-sm4,-avx512ifma,-avx512pf,-tsxldtrk,-ptwrite,-widekl,-sm3,-xsavec,-avx10.1-512,-avx512vpopcntdq,-avx512vp2intersect,-avx512cd,-avxvnniint8,-avx512er,-amx-int8,-kl,-avx10.1-256,-sha512,-avxvnni,-hreset,-movdiri,-serialize,-vpclmulqdq,-avx512vl,-uintr,-clflushopt,-raoint,-cmpccxadd,-amx-tile,-gfni,-avxvnniint16,-amx-fp16,-avx512f,-amx-bf16,-avx512bf16,-avx512vnni,-avx512bw,-pku,-clzero,-mwaitx,-lwp,-sha,-movdir64b,-wbnoinvd,-enqcmd,-prefetchwt1,-avxneconvert,-tbm,-pconfig,-amx-complex,-avxifma,-avx512bitalg,-rdpru,-clwb,-avx512vbmi2,-prefetchi,-rdpid,-fma4,-avx512vbmi,-shstk,-vaes,-waitpkg,-sgx,-avx512dq,-sse4a"
* "cascadelake"
  "+prfchw,+avx,+aes,+sahf,+pclmul,+crc32,+xsaves,+sse4.1,+xsave,+sse4.2,+invpcid,+64bit,+xsavec,+cmov,+avx512cd,+movbe,+evex512,+adx,+avx2,+avx512vl,+clflushopt,+bmi,+sse,+xsaveopt,+rdrnd,+avx512f,+avx512vnni,+cx8,+avx512bw,+sse3,+pku,+fsgsbase,+lzcnt,+ssse3,+cx16,+bmi2,+fma,+popcnt,+f16c,+clwb,+mmx,+sse2,+rdseed,+fxsr,+avx512dq,-cldemote,-xop,-avx512fp16,-usermsr,-sm4,-avx512ifma,-avx512pf,-tsxldtrk,-ptwrite,-widekl,-sm3,-avx10.1-512,-avx512vpopcntdq,-avx512vp2intersect,-avxvnniint8,-avx512er,-amx-int8,-kl,-avx10.1-256,-avxvnni,-rtm,-hreset,-movdiri,-serialize,-sha512,-vpclmulqdq,-uintr,-raoint,-cmpccxadd,-amx-tile,-gfni,-avxvnniint16,-amx-fp16,-amx-bf16,-avx512bf16,-clzero,-mwaitx,-lwp,-sha,-movdir64b,-wbnoinvd,-enqcmd,-prefetchwt1,-avxneconvert,-tbm,-pconfig,-amx-complex,-avxifma,-avx512bitalg,-rdpru,-avx512vbmi2,-prefetchi,-rdpid,-fma4,-avx512vbmi,-shstk,-vaes,-waitpkg,-sgx,-sse4a"
* "skylake-avx512"
  "+cx16,+sahf,+crc32,+prfchw,+bmi2,+fsgsbase,+popcnt,+aes,+xsaves,+clwb,+avx512f,+xsavec,+pku,+mmx,+rdseed,+avx512bw,+clflushopt,+xsave,+64bit,+avx512vl,+invpcid,+avx512cd,+avx,+cx8,+fma,+bmi,+rdrnd,+sse4.1,+sse4.2,+avx2,+fxsr,+sse,+lzcnt,+pclmul,+f16c,+ssse3,+cmov,+movbe,+xsaveopt,+avx512dq,+sse2,+adx,+sse3,-avx512pf,-tsxldtrk,-tbm,-avx512ifma,-sha,-fma4,-vpclmulqdq,-cldemote,-avx512bf16,-amx-tile,-raoint,-uintr,-gfni,-ptwrite,-avx512bitalg,-movdiri,-widekl,-avx512er,-avxvnni,-avx512fp16,-avx512vnni,-amx-bf16,-avxvnniint8,-avx512vpopcntdq,-pconfig,-cmpccxadd,-clzero,-amx-fp16,-lwp,-rdpid,-xop,-waitpkg,-prefetchi,-kl,-movdir64b,-sse4a,-avxneconvert,-avx512vbmi2,-serialize,-hreset,-vaes,-amx-int8,-rtm,-enqcmd,-mwaitx,-wbnoinvd,-rdpru,-avxifma,-sgx,-prefetchwt1,-avx512vbmi,-shstk,-avx512vp2intersect"
* "skylake-avx512"
  "+prfchw,+avx,+aes,+sahf,+pclmul,+crc32,+sse4.1,+xsave,+sse4.2,+64bit,+cmov,+movbe,+sse,+rdrnd,+cx8,+sse3,+fsgsbase,+ssse3,+cx16,+fma,+popcnt,+f16c,+mmx,+sse2,+fxsr,-cldemote,-xop,-xsaves,-avx512fp16,-usermsr,-sm4,-avx512ifma,-avx512pf,-tsxldtrk,-ptwrite,-widekl,-sm3,-invpcid,-xsavec,-avx10.1-512,-avx512vpopcntdq,-avx512vp2intersect,-avx512cd,-avxvnniint8,-avx512er,-amx-int8,-kl,-avx10.1-256,-sha512,-avxvnni,-rtm,-adx,-avx2,-hreset,-movdiri,-serialize,-vpclmulqdq,-avx512vl,-uintr,-clflushopt,-raoint,-cmpccxadd,-bmi,-amx-tile,-gfni,-avxvnniint16,-amx-fp16,-xsaveopt,-avx512f,-amx-bf16,-avx512bf16,-avx512vnni,-avx512bw,-pku,-clzero,-mwaitx,-lwp,-lzcnt,-sha,-movdir64b,-wbnoinvd,-enqcmd,-prefetchwt1,-avxneconvert,-tbm,-pconfig,-amx-complex,-bmi2,-avxifma,-avx512bitalg,-rdpru,-clwb,-rdseed,-avx512vbmi2,-prefetchi,-rdpid,-fma4,-avx512vbmi,-shstk,-vaes,-waitpkg,-sgx,-avx512dq,-sse4a"
* "icelake-server"
  "+prfchw,+avx,+aes,+sahf,+pclmul,+crc32,+sse4.1,+xsave,+sse4.2,+64bit,+cmov,+movbe,+sse,+rdrnd,+cx8,+sse3,+fsgsbase,+ssse3,+cx16,+fma,+popcnt,+f16c,+mmx,+sse2,+fxsr,-cldemote,-xop,-xsaves,-avx512fp16,-usermsr,-sm4,-avx512ifma,-avx512pf,-tsxldtrk,-ptwrite,-widekl,-sm3,-invpcid,-xsavec,-avx10.1-512,-avx512vpopcntdq,-avx512vp2intersect,-avx512cd,-avxvnniint8,-avx512er,-amx-int8,-kl,-avx10.1-256,-sha512,-avxvnni,-rtm,-adx,-avx2,-hreset,-movdiri,-serialize,-vpclmulqdq,-avx512vl,-uintr,-clflushopt,-raoint,-cmpccxadd,-bmi,-amx-tile,-gfni,-avxvnniint16,-amx-fp16,-xsaveopt,-avx512f,-amx-bf16,-avx512bf16,-avx512vnni,-avx512bw,-pku,-clzero,-mwaitx,-lwp,-lzcnt,-sha,-movdir64b,-wbnoinvd,-enqcmd,-prefetchwt1,-avxneconvert,-tbm,-pconfig,-amx-complex,-bmi2,-avxifma,-avx512bitalg,-rdpru,-clwb,-rdseed,-avx512vbmi2,-prefetchi,-rdpid,-fma4,-avx512vbmi,-shstk,-vaes,-waitpkg,-sgx,-avx512dq,-sse4a"  
* "sandybridge"
  "+avx,+aes,+sahf,+pclmul,+crc32,+sse4.1,+xsave,+sse4.2,+64bit,+cmov,+sse,+xsaveopt,+cx8,+sse3,+ssse3,+cx16,+popcnt,+mmx,+sse2,+fxsr,-prfchw,-cldemote,-xop,-xsaves,-avx512fp16,-usermsr,-sm4,-avx512ifma,-avx512pf,-tsxldtrk,-ptwrite,-widekl,-sm3,-invpcid,-xsavec,-avx10.1-512,-avx512vpopcntdq,-avx512vp2intersect,-avx512cd,-movbe,-avxvnniint8,-avx512er,-amx-int8,-kl,-avx10.1-256,-sha512,-avxvnni,-rtm,-adx,-avx2,-hreset,-movdiri,-serialize,-vpclmulqdq,-avx512vl,-uintr,-clflushopt,-raoint,-cmpccxadd,-bmi,-amx-tile,-gfni,-avxvnniint16,-amx-fp16,-rdrnd,-avx512f,-amx-bf16,-avx512bf16,-avx512vnni,-avx512bw,-pku,-fsgsbase,-clzero,-mwaitx,-lwp,-lzcnt,-sha,-movdir64b,-wbnoinvd,-enqcmd,-prefetchwt1,-avxneconvert,-tbm,-pconfig,-amx-complex,-bmi2,-fma,-avxifma,-f16c,-avx512bitalg,-rdpru,-clwb,-rdseed,-avx512vbmi2,-prefetchi,-rdpid,-fma4,-avx512vbmi,-shstk,-vaes,-waitpkg,-sgx,-avx512dq,-sse4a"

```

topolarity replied:
That's a good point. We probably need to tighten that up.


I'm curious to have @gbaraldi 's input on this

gbaraldi replied:
This is just a huge mess. some VMs hide features for example

gbaraldi replied:
I guess to be clear, the mess I mean is as a whole. I think this is mergeable without too many issues but I wish there was something nicer that LLVM exposed, but there isn't currently

topolarity replied:
ref: https://github.com/JuliaLang/julia/pull/57049

tz-lom commented on src/HostCPUFeatures.jl:
```diff
@@ -37,19 +38,26 @@ unwrap(::StaticSymbol{S}) where {S} = S
   reset_extra_features!()
 end
 const BASELINE_CPU_NAME = get_cpu_name()
+const allow_eval = @load_preference("allow_runtime_invalidation", false)
```
Why not to check that `--trim` is enabled in JLOptions ?

everyone who will use it with trim will have to find out this culprit themself

topolarity replied:
The main thinking is that this preference is usable by more than just JuliaC.jl, since some users may wish to be opt-in to error / warn on invalidation storms from this package. JuliaC.jl can set this preference automatically, so the end-user experience is the same.

Also technically checking `JLOptions` at pre-compilation time won't detect `--trim` properly, but something like https://github.com/JuliaLang/JuliaC.jl/issues/31 would work

tz-lom commented:
As a source of the problem I understand the approach, but it actually downgrades package `HostCPUFeatures` to `MaybeSomeOtherHostCPUFeatures`, while what we really need is to use some `CompileTargetCPUFeatures` in `LoopVectorization` etc.

Pragmatically it is "easier" upgrade standard interface of `HostCPUFeatures` to return features for the compile target and provide API to retrieve real Host features if that what user really wants (like printing statistics on which CPU we are running now)

topolarity commented:
> As a source of the problem I understand the approach, but it actually downgrades package HostCPUFeatures to MaybeSomeOtherHostCPUFeatures, while what we really need is to use some CompileTargetCPUFeatures in LoopVectorization etc.

That's already true if `"native"` is missing from your `JULIA_CPU_TARGET` string though - I agree that it might be a good idea to require that we are specifically operating in the non-`native` case though (where we already under-approximate feature flags)

topolarity committed:
Add `build_cpu_target` global to opt-in to generic CPU at precompile-time

We already support this dynamically, but for some reason it wasn't
possible to get this configuration at precompilation time. Also adds
a "freeze_cpu_target" preference to enable this behavior by default.

gbaraldi commented on src/cpu_info.jl:
```diff
@@ -39,17 +39,22 @@ function set_features!()
         if _has_feature(feature) ≠ has
-            @debug "Defining $(has ? "presence" : "absense") of feature $feature."
-            set_feature(feature, has)
+            if allow_eval
```
This code doesn't execute if there's a preference right?

topolarity replied:
Depends on which preference you mean

Are you saying you don't think this guard is necessary?

oscardssmith commented:
Looks like the Julia 1.5 CI is failing. Time to bump minimum version to LTS?

oscardssmith commented:
Other than that, LGTM.

oscardssmith commented:
@tz-lom are you good with this in current state? If so, I'll merge and release.

tz-lom commented:
> @tz-lom are you good with this in current state? If so, I'll merge and release.

@oscardssmith , thanks, it looks ok.
I think you need to fix test matrix to solve failing tests, and it would be nice if you put description of preferences in documentation, because now you have to dig through the code to find out this feature.

topolarity committed:
Bump minimum Julia version to 1.10

topolarity committed:
Add preferences documentation

topolarity commented:
@tz-lom Anything else you need from me on this?

tz-lom commented:
@topolarity , no that is enough, thank you very much for your work

tz-lom commented:
@topolarity hi Cody, I think we did whoopsie here
I think version was not bumped, and I see that link to documentation from the badge in README is not working (wrong link?)

also check open issue https://github.com/JuliaSIMD/HostCPUFeatures.jl/issues/15 , maybe it can be closed?

oscardssmith commented:
I'll bump and release.

oscardssmith commented:
@topolarity you missed a compat entry on Preferences.jl
